### PR TITLE
Add 'user notes' functionality

### LINKED ISSUE #135: Suggestion: User notes
A system to create notes on users and recall them either on command or automatically on a certain trigger eg. an infraction being added or a modmail thread being opened.

It would consist of a /note command, with the ability to create, query and remove notes on a certain user, the bot sending a message to a channel (and mentioning the involved mod) when a user is warned, and the bot sending a message to the thread when a new modmail thread is created. This system would likely replace modmail's built-in note system to prevent fragmentation.

Config options for each note would probably include setting it to either remove itself after triggering once or be persistent, and being able to set a note to trigger either only for yourself or any mod that warns a user.

### PR DESCRIPTION
**Pull request by FloatingMilkshake**
This PR closes #135.

Adds the ability to set notes on users. These notes are shown only to moderators, and can be queried manually or shown automatically when the target user is warned or opens a modmail thread. Notes can also be set to show only one time, after which they will be removed. The `/note` command offers options to add, edit, delete, list, and show details for notes. 

Commands that require referencing a note (e.g. the target note in `/note details`) support autocomplete, just like `/warndetails` & similar.

image

Note embeds are very similar to warning embeds. Note commands are all ephemeral by default, but `/note list` and `/note details` can be shown publicly in chat for ease of reference.

image
image

When a modmail thread is opened for a user who has notes set to 'Show on Modmail', Cliptok will send an embed with only the relevant notes into the thread channel, similar to how it already sends warning & mute information if the user is muted.

image

When a user with notes set to 'Show on Warn' is issued a warning, Cliptok will send an alert into #investigations pinging the moderator responsible for the warning. Relevant notes will be displayed, but only if either A) they are set to 'Show all Mods', or B) the moderator responsible for the warning was also responsible for the note(s).

image

'Show all Mods' only applies to warning alerts. Notes set to 'Show on Modmail' will be displayed in modmail threads regardless of the 'Show all Mods' setting. 

Notes are stored in Redis directly next to warnings—they are _**not**_ stored under a separate hash like `userNotes`. They use the same ID namespace as warnings. Notes and warnings both have a new `WarningType` property to help distinguish between them. This type can either be `Warning` or `Note`. Older warnings without a type are treated as having the type `Warning`.

### CODE REVIEW

File: Helpers/UserNoteHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                    null,
+                    await LykosAvatarMethods.UserOrMemberAvatarURL(user, Program.homeGuild, "png")
+                )
+                .AddField("Note ID", StringHelpers.Pad(note.NoteId))
+                .AddField("Show on Modmail", note.ShowOnModmail ? "Yes" : "No")
+                .AddField("Show on Warn", note.ShowOnWarn ? "Yes" : "No")
+                .AddField("Show all Mods", note.ShowAllMods ? "Yes" : "No")
+                .AddField("Show Once", note.ShowOnce ? "Yes" : "No")
+                .AddField("Responsible moderator", $"<@{note.ModUserId}>")
+                .AddField("Time", $"<t:{TimeHelpers.ToUnixTimestamp(note.Timestamp)}:f>");
```
Comment (Erisa) [inline Helpers/UserNoteHelpers.cs:92]:
```suggestion
                .AddField("Note ID", StringHelpers.Pad(note.NoteId), true)
                .AddField("Show on Modmail", note.ShowOnModmail ? "Yes" : "No", true)
                .AddField("Show on Warn", note.ShowOnWarn ? "Yes" : "No", true)
                .AddField("Show all Mods", note.ShowAllMods ? "Yes" : "No", true)
                .AddField("Show Once", note.ShowOnce ? "Yes" : "No", true)
                .AddField("Responsible moderator", $"<@{note.ModUserId}>", true)
                .AddField("Time", $"<t:{TimeHelpers.ToUnixTimestamp(note.Timestamp)}:f>", true);
```

File: Events/MessageEvent.cs [hunk 2]
```diff
@@ -119,11 +119,37 @@ public static async Task MessageHandlerAsync(DiscordClient client, DiscordMessag
...
+                    
+                    // Filter to notes set to notify on modmail
+                    var notesToNotify = notes.Where(x => x.Value.ShowOnModmail).ToDictionary(x => x.Key, x => x.Value);
+                    
+                    // If there are notes, build embed and add to message
+                    if (notesToNotify.Count != 0)
+                        memberWarnInfo.AddEmbed(await UserNoteHelpers.GenerateUserNotesEmbedAsync(modmailMember, notesToUse: notesToNotify));
+                    
+                    // If message was built (if user is muted OR if user has notes to show on modmail), send it
+                    if (memberWarnInfo.Embeds.Count != 0) // todo: this is probably not the best way to check this?
```
Comment (Erisa) [inline Events/MessageEvent.cs:147]: > this is probably not the best way to check this?

I think this actually makes perfectly sense to me.

Comment (FloatingMilkshake) [reply Events/MessageEvent.cs:147]: Okay! Will leave it as-is then.

**Review by Erisa — CHANGES_REQUESTED** [review]
Can we log all note events to the mod log, please?

**Comment by FloatingMilkshake** [comment]
> Can we log all note events to the mod log, please?

Absolutely! Done in 3d27ab3

**Review by Erisa — APPROVED** [review]
lgtm

**Pull request merged**